### ISSUE TITLE
Conduits loading neighboring chunks even if not doing work

### ISSUE THREAD

**Issue by RoyCurtis** [SleepyTrousers/EnderIO#2673]
It appears that any tile entity of type "EnderIO:tile.blockConduitBundle" in world will forcefully load neighbouring chunks, causing a cascading effect.

This is regardless of if the conduits are doing any work or not (e.g. transferring items, fluid, energy, reciving a redstone signal, etc.).

This affects: Redstone, Insluated Redstone, Fluid (any tier), Energy (any tier) and item conduits, as well as conduit switches and conduit facades.

# Details

## Discovered on

* Direwolf20 1.7.10 1.5.0 and 1.6.1 live server
* Forge 10.13.4.1448
* [EnderIO 2.2.8.381](http://minecraft.curseforge.com/mc-mods/64578-ender-io/files/2238375)

# Reproduction

Instead of detailing a specific reproduction, I will demonstrate some example cases of this bug manifesting and explain why and how they manifest.

## Testing environment

* Forge 1.7.10 10.13.4.1448
* [EnderIO](http://ci.tterrag.com/job/EnderIO/) 2.3.0.403_beta
* [EnderCore](http://ci.tterrag.com/job/EnderCore/) 0.0.1.13_alpha
* Chickenbones mods
  * [CodeChickenCore](http://chickenbones.net/maven/codechicken/CodeChickenCore/1.7.10-1.0.7.46/) 1.0.7.46
  * [ChickenChunks](http://chickenbones.net/maven/codechicken/ChickenChunks/1.7.10-1.3.4.19/) 1.3.4.19
  * [NotEnoughItems](http://chickenbones.net/maven/codechicken/NotEnoughItems/1.7.10-1.0.5.111/) 1.0.5.111

## Setup

*Updated July 28th to use cobblestone flatland*

For each of these use cases, I perform the following the steps to demonstrate the chunkloading. For each case, ChickenChunks is used to forcefully keep a chunk loaded, which this bug relies on. Note that I use orange and blue wool only to show the chunk borders.

1. Create a flatland creative world with:
  * Cheats enabled
  * Superflat generation with the preset `2;7,4;1;`
  * Structures off
2. Execute `/gamerule doMobSpawning false`
3. Execute `/tp 0 8 0` and then `/setworldspawn` (keeps spawn chunks away from testing areas)
4. Execute `/tp 1000 8 1000`
5. [Place a ChickenChunks spot loader in the middle of a chunk, to keep it loaded](http://i.imgur.com/nMbQeQD.jpg)
6. Execute `/tp 0 8 0`
7. Press ESC to open the main menu and force a chunk purge
8. Resume game and execute `/chunkloaders`. A new window will open under Minecraft
9. [Note where the chunkloader is located](http://i.imgur.com/IDiSRlh.png) and observe how no neighbouring chunks are loaded

## Test cases

*Note: I have not used Redstone Conduits for these tests because of a [pending bug report](https://github.com/SleepyTrousers/EnderIO/issues/2672)*

### Case 1: Dead conduits causing chunk loads

* Screenshot: http://i.imgur.com/jBP5p5q.jpg
  1. Energy conduit with no energy flowing and no connections, passing the north chunk boundary
  2. Fluid conduit with no fluid flowing and no connections, passing the east chunk boundary
  3. Item conduit with no item flowing and no connections, passing the south chunk boundary
  4. Conduit switch with no additional conduits and no connections, passing the west chunk boundary
* Result: [4 chunks are kept loaded](http://i.imgur.com/8KSXt8t.png)

### Case 2: Extended chunk loads by dead conduits passing further borders

* Screenshot: http://i.imgur.com/gmbUczv.jpg
* Result: 8 chunks are kept loaded

### Case 3: Cascading chunk load by disconnected distant conduits

This case is subjectively the most telling of potential issues with this bug, where unrelated and distant conduits cause chunkloading even if they are doing no work.

* Screenshot: http://i.imgur.com/0K81Sav.jpg
  1. Chunk forcefully loaded with chunkloader
  2. Energy conduit with no energy flowing and no connections, passing the south chunk boundary
  3. Energy conduit with no energy flowing and no connections, passing another south chunk boundary, ad nauseum
* Result: [5 chunks are kept loaded](http://i.imgur.com/qnUjN46.png)

### Case 4: Cascading chunk load by conduits touching chunk boundaries

* Screenshot: http://i.imgur.com/MPaUibd.jpg
  1. Chunk forcefully loaded with chunkloader
  2. Energy conduit with no energy flowing and no connections, passing the south chunk boundary
  3. Energy conduit with no energy flowing and no connections, *touching only the south chunk boundary* instead of crossing it
  4. Energy conduit with no energy flowing and no connections, *touching only the east chunk boundary* instead of crossing it
* Result: [5 chunks are kept loaded](http://i.imgur.com/qnUjN46.png). It is interesting to note that these chunks load "slower" (by a few miliseconds) than other cases

### Case 5: Cascading chunk load by facades

* Screenshot: http://i.imgur.com/y2aNPcO.jpg
  1. Conduit facade with no conduit bundles inside, passing the east chunk boundary
  2. Energy conduit with no energy flowing and no connections, passing the south chunk boundary
  3. Conduit facade with no conduit bundles inside, passing another south chunk boundary
  4. Conduit facade with no conduit bundles inside, *touching only the east chunk boundary* instead of crossing it
* Result: [5 chunks are kept loaded](http://i.imgur.com/qnUjN46.png). Note that the facade passing the east chunk boundary from the chunkloader does not load that chunk. It appears facades only chunkload as part of a cascade.

# Notes

Please excuse the large bug report. I believe that this is a serious issue that explains a lot of the performance issues we have been experiencing on our server, in regards to chunkloaders repeatedly loading and unloading.

I realize there are two pending [bug reports of a similar nature](https://github.com/SleepyTrousers/EnderIO/issues/2672), which have [code fixes](https://github.com/SleepyTrousers/EnderIO/issues/2671) that may fix this issue. I wanted to report this issue for the benefit of server owners as well as the developers.

I think also these previous reports are related:

* [Conduit lag #1959](https://github.com/SleepyTrousers/EnderIO/issues/1959)
* [Stall reports #2648](https://github.com/SleepyTrousers/EnderIO/issues/2648)

**Comment by tterrag1098** [comment]
Simply placing a spotloader in the chunk and nothing else results in this behavior, where the chunks will periodically unload and load again.

![](http://puu.sh/jdH4X.png)

I'm not sure how I can test reliably with this happening.

**Comment by RoyCurtis** [comment]
That isn't good. I am seeing the same thing; sorry I missed this the first time. I will try to figure out what is going on.

**Comment by RoyCurtis** [comment]
OK; for an unknown reason, chunks randomly load around spotloaders *whilst there is sunlight*. I do not know if that is a ChickenChunks, Forge or Vanilla issue. The chunk I picked for the test case happened to not have that issue by chance.

Here's another test case;

* Add CoFH Core as a mod
* Execute `/tp 200 6 600`
* [Set up a spot loader and random conduits crossing chunk boundaries in a figure of 8 formation](http://i.imgur.com/LR88IxO.jpg)

To make sure other factors are not affecting the test:

* Prevent spawning with `/gamerule doMobSpawning false`
* Execute `/cofh killall a` - repeat for e, i, o and u to kill all entities
* Stop time by doing `/gamerule doDaylightCycle false`
* Set time to day with `/time set day`

[Whilst the problem you encountered happens here](http://i.imgur.com/e4w09V2.png) (e.g. the chunk north and north west of the spotloader), it is only happening to chunks surrounding the spotloader. Chunks loaded by the conduits are not loading neighboring chunks themselves.

If I [move the spotloader one chunk to the north so that it is not in a chunk with conduits](http://i.imgur.com/euNmBD5.jpg), it will eventually load the south chunk and [consequently all the connected chunks](http://vanderprot.gamealition.com/img/2015-07-27_10-17-37.mp4) (16 second video).

Interestingly, when there is no sunlight, the loading behavior changes. After doing `/time set 18000` ("night" apparently still counts as sunlight), spotloaders no longer load their surrounding chunks at random. When the spotloader is put back into the chunk with the conduits, [the chunks connected by conduits rapidly load and unload but the entire figure of 8 never loads fully](http://vanderprot.gamealition.com/img/2015-07-27_10-32-56.mp4).

----

To summarize, try the test cases whilst the time is set to 18000 ticks. The cascading behaviors will no longer be present but I think the rapid loading and unloading of chunks is still representative of an issue with the conduits. I'll keep testing other things about this issue during the day.

**Comment by RoyCurtis** [comment]
Thanks to a back and forth between @utoc and I, it's since been found that the grass was causing chunks to randomly load around the chunkloader.

I will update the original report and test cases with a better reproduction

**Comment by tterrag1098** [comment]
The above commit seems to alleviate your issues, however (even though there were none I could see) I am unsure of the ramifications it could have in other areas. Further testing is needed.

**Comment by VapourDrive** [comment]
Isn't "/cofh killall *" a quicker way to remove all entities?
This is like the posterchild of bug reports though!

**Comment by tterrag1098** [comment]
@VapourDrive just killall only removes hostile mobs, unfortunately.

**Comment by RoyCurtis** [comment]
I will try testing over the next 24 hours, in both the original testing environment and on a Direwolf20 1.6.1 (with the updated EnderIO) testing server. Thanks!

**Comment by VapourDrive** [comment]
I think the command may have changed: 
![cofhkillall](https://cloud.githubusercontent.com/assets/5392006/9024629/803df9e0-38a7-11e5-8243-d07c2ce5cb65.png)
(as taken from the cofh website)

**Comment by RoyCurtis** [comment]
Tested the latest dev builds; issue is still reproducible with naked redstone conduits and conduit switches. I used the same testing environment with two differences:

* [EnderCore 0.0.1.15_alpha](http://ci.tterrag.com/job/EnderCore/15/)
* [EnderIO 2.3.0.405_beta](http://ci.tterrag.com/job/EnderIO/405/)

I also used a separate development environment to try and figure out what is going on at the source level, [by breakpointing `ChunkEvent.Load`](http://i.imgur.com/Lvms1mV.png). *Please excuse any inaccurate assumptions I make about EnderIO's internals; feel free to correct me*

## Case 1: Dead conduits no longer causing chunkloads

This case is just a demonstration of the issues fixed by commit https://github.com/SleepyTrousers/EnderIO/commit/f6d2de4c4361818e476b5155322b8c61d07aeea9

* Screenshot: http://i.imgur.com/vTt63cy.jpg
  1. Energy conduit with no energy flowing and no connections, passing the north chunk boundary
  2. Insulated redstone conduit with no signal and no connections, passing the east chunk boundary
  3. Fluid conduit with no fluid flowing and no connections, passing the south chunk boundary
  4. Item conduit with no items flowing and no connections, passing the west chunk boundary
* Result: **Case resolved**; [no chunks are kept loaded](http://i.imgur.com/Mi1fgwg.png)

## Case 2: Redstone conduits and conduit switches

* Screenshot: http://i.imgur.com/w7UycUj.jpg
  1. Conduit switches with no signal or connections crossing the north and east chunk boundaries
  2. Naked redstone conduits with no signal or connections crossing the south and west chunk boundaries
* Result: [4 chunks are kept loaded](http://i.imgur.com/hey0W7B.png)

[By breakpointing this chunk load in the dev environment](http://i.imgur.com/Ujfw1wc.png), it appears [a call to `world.getIndirectPowerLevelTo`](https://github.com/SleepyTrousers/EnderIO/blob/master/src/main/java/crazypants/enderio/conduit/redstone/RedstoneConduit.java#L222) causes these conduits to load other chunks.

I think what is happening is that each redstone/switch conduit considers every side a network connection, even [if nothing is connected](http://i.imgur.com/zcdY10o.png). By contrast, insulated redstone conduits only check explicit connections, so they no longer trigger this bug [unless they are connected to a signal on either side](http://i.imgur.com/BpdhNtC.png).

## Case 3: Cascading chunk load by disconnected distant conduits

If a chunk is loaded by case 2, subsequent chunks are loaded in cascade by all other conduit types if they pass chunk boundaries. *This case may require a few presses of ESC or `/save-all` to properly trigger*

* Screenshot: http://i.imgur.com/MgKpCol.jpg
  1. Naked redstone conduit with no signal and no connections, passing the south chunk boundary
  2. Fluid conduit with no fluid flowing and no connections, passing the next south chunk boundary
  3. Energy conduit with no energy flowing and no connections, passing the next east chunk boundary
  4. Item conduit with no items flowing and no connections, passing the next east chunk boundary
  5. Fluid conduit with no fluid flowing and no connections, passing the next north chunk boundary
* Result: [5 chunks are kept loaded](http://i.imgur.com/ZNAvmw7.png)

By breakpointing the cascade chunk loads in the dev environment, [it appears that the disconnected conduits are marked "dirty" ad nauseum](http://i.imgur.com/t0ssVr6.png). This is despite being disconnected from the naked redstone conduit. This is what I think is happening:

1. The naked redstone conduit crossing the south boundary from the spotloaded chunk, causes the south chunk to load (see case 2)
2. The naked redstone conduit marks the fluid conduit passing the next south boundary as "dirty"
3. The fluid conduits are "updated" and consequently load both chunks it crosses, then marks the next pair of conduits as "dirty"
4. Repeat ad nauseum

I am unable to figure out where or when the conduits are being marked dirty.

-----

I will wait to test these fixes with a Direwolf20 1.6.1 server, to ensure there are no ramifications.

@VapourDrive: Thank you; I did not know the killall command could do that. I have updated my comment.

**Comment by tterrag1098** [comment]
I never specifically tested the redstone conduits, there is probably some more unsafe code in there. I will investigate tonight hopefully.
